src/main/java/org/javarosa/xpath/expr/XPathEncryptStringFunc.java

```diff
@@ -0,0 +1,84 @@
+package org.javarosa.xpath.expr;
+
+
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.instance.DataInstance;
+import org.javarosa.xpath.XPathException;
+import org.javarosa.xpath.parser.XPathSyntaxException;
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.SecureRandom;
+import java.util.Base64;
+
+public class XPathEncryptStringFunc extends XPathFuncExpr {
+    public static final String NAME = "encrypt-string";
+    private static final int EXPECTED_ARG_COUNT = 3;
+
+    public XPathEncryptStringFunc() {
+        name = NAME;
+        expectedArgCount = EXPECTED_ARG_COUNT;
+    }
+
+    public XPathEncryptStringFunc(XPathExpression[] args) throws XPathSyntaxException {
+        super(NAME, args, EXPECTED_ARG_COUNT, true);
+    }
+
+    @Override
+    public Object evalBody(DataInstance model, EvaluationContext evalContext, Object[] evaluatedArgs) {
+        return encryptString(evaluatedArgs[0], evaluatedArgs[1], evaluatedArgs[2]);
+    }
+
+    /**
+     * Encrypt a message with the given algorithm and key.
+     *
+     * @param message    a message to be encrypted
+     * @param key        the key used for encryption
+     * @param algorithm  the encryption algorithm to use
+     */
+    private static String encryptString(Object o1, Object o2, Object o3) {
+        String message = FunctionUtils.toString(o1);
+        String key = FunctionUtils.toString(o2);
+        String algorithm = FunctionUtils.toString(o3);
+
+        final String ENCRYPT_ALGO = "AES/GCM/NoPadding";
+        final int TAG_LENGTH_BIT = 128;
+        final int IV_LENGTH_BYTE = 12;
+        final int KEY_LENGTH_BIT = 256;
+
+        if (!algorithm.equals("AES")) {
+            throw new XPathException("Unknown algorithm \"" + algorithm +
+                                     "\" for " + NAME);
+        }
+
+        Base64.Decoder keyDecoder = Base64.getUrlDecoder();
+        byte[] keyBytes = keyDecoder.decode(key);
+        if (8 * keyBytes.length != KEY_LENGTH_BIT) {
+            throw new XPathException("Key should be " + KEY_LENGTH_BIT +
+                                     " bits long, not " + 8 * keyBytes.length);
+        }
+        SecretKey secret = new SecretKeySpec(keyBytes, "AES");
+
+        byte[] iv = new byte[IV_LENGTH_BYTE];
+        new SecureRandom().nextBytes(iv);
+
+        try {
+            Cipher cipher = Cipher.getInstance(ENCRYPT_ALGO);
+            cipher.init(Cipher.ENCRYPT_MODE, secret, new GCMParameterSpec(TAG_LENGTH_BIT, iv));
```

Comment:
Hm, this might cause issues on Android, because android didn't support GCM schemes until API 19 
https://developer.android.com/reference/javax/crypto/spec/GCMParameterSpec.

We still have devices running older versions, which means that this code probably won't actually even compile for us on Android. 

Is there a way to get the GCM from a polymorphic factory indirection callout like the Cipher instance factory? It's fine for the code to not *work* on Android 4.1, but I think the class signature needs to compile.

I'm trying to find an easy way to make that testable, but it looks like our cross-request CI builder only works with branches that are in our repo directly.

Reply:
Some good news here. At encryption time, if we just specify the algorithm as "AES/GCM/NoPadding", it will use the same parameters we were specifying manually - 12 byte initialization vector, generated internally, 128-bit tag. So we can make that code compatible back to [API level 10](https://developer.android.com/guide/topics/security/cryptography#SupportedCipher), I believe.

In the test code, however, I haven't been able to get rid of the GCMParameterSpec. In the past, it seems you could use an IvParameterSpec, but that is causing an exception for me using the crypto provider in my JDK. Does the test code need to compile for Android 4.1 too?

I have not tracked down any factory to make the GCM parameters yet.